docs/testing.md

```diff
@@ -0,0 +1,221 @@
+
+### Testing User Interactions
+
+Aside from rendering some UI, your components handle events like `onChangeText` for `TextInput` or `onPress` for `Button`. They may also contain other functions and event callbacks. Consider the following example:
+
+```js
+class GroceryShoppingListCreator extends React.Component {
+  render() {
+    return (
+      <>
+        <TextInput
+          placeholder="Enter grocery item"
+          onChangeText={this.setGroceryItem}
+        />
+        <Button
+          title="Add the item to list"
+          onPress={this.props.addNewItemToShoppingList}
+        />
+      </>
+    );
+  }
+}
+```
+
+When testing user interactions, test the component from the user perspective: you could write your tests to call `addNewItemToShoppingList` directly, and while such test works, it's very close to the implementation details of your component and would break if you refactor the component.
+
+To counter for that, component testing libraries such as [`react-native-testing-library`](https://github.com/callstack/react-native-testing-library), offer `fireEvent` apis that simulate a user interacting with the component. There are apis that allow to simulate entering text into textinput, tapping buttons and more. An example of how you may fire text change event using `react-native-testing-library`:
+
+```js
+test('updates grocery input', () => {
+  fireEvent.changeText(getByPlaceholder('Enter grocery item'), 'banana');
+});
+```
+
+The first parameter is a query function that returns a `ReactTestInstance` and the second parameter is the new text that user entered.
+
+This way, we're not testing what some function is doing - we're testing what happens when a user changes the text in the `TextInput`!
+
+Now that we know how to fire an event, let's take a look at verifying that it caused the expected change in your component: for example, that a certain text is rendered in a `GroceryShoppingList`. Aforementioned component testing libraries expose `query` apis for this purpose. An example may look like this:
+
+```js
+test('given empty GroceryShoppingList, user can add an item to it', () => {
+  const {getByTestId, getByText, getAllByText} = render(
+    <GroceryShoppingList />,
+  );
+
+  fireEvent.changeText(getByPlaceholder('Enter grocery item'), 'banana');
+  fireEvent.press(getByText('Add the item to list'));
+  const bananaElements = getAllByText('banana');
+  expect(bananaElements).toHaveLength(1); // expect 'banana' to be on the list
+});
+```
+
+### Testing Rendered Output
+
+[Snapshot testing](https://jestjs.io/docs/en/snapshot-testing) lets you verify how component is rendering. A snapshot is a textual representation of your component’s render output, and may look like this:
```

Comment:
What do you think about this phrasing?
```suggestion
[Snapshot testing](https://jestjs.io/docs/en/snapshot-testing) is a special kind of testing enabled by Jest. It is a very powerful and low-level tool, so special attention is advised when using it. 

> In this guide we will speak about "component snapshot", which is a snapshot specialized to render React component tree objects using a JSX-like string, powered by a [custom React serializer](https://jestjs.io/docs/en/configuration#snapshotserializers-arraystring) that lets Jest translate JS objects to string that's human-readable. It's built into Jest and used by default so you don't need to worry about it.

A component snapshot is a textual representation of your component’s render output _generated_ during a test run, and may look like this:
```

Reply:
@thymikee I took the suggestion but not sure about it upon re-reading it: If I simplify what's written there, it sounds like "component snapshot is a snapshot specialized to render React components using a string",  which sounds a little odd.

here's the text for reference:

> In this guide we will speak about "component snapshot", which is a snapshot specialized to render React component tree objects using a JSX-like string, powered by a custom React serializer that lets Jest translate JS objects to string that's human-readable. 

How about this?

In this guide we will speak about "component snapshot", which is a JSX-like string, created by a custom React serializer that lets Jest translate React component trees to string that's human-readable.

Reply:
Sounds good 👍